### PR TITLE
Adding docs

### PR DESCRIPTION
**Pull request by Vinitha-balachandran**
To add doc files

### CODE REVIEW

**Comment by alecandido** [comment]
TODO: point this PR to the branch of #24 (I can't do immediately, because there is no new commit wrt that one)

Motivation: having a meaningful diff (i.e. a meaningful view in the *Files changed* tab)

**Comment by Vinitha-balachandran** [comment]
> TODO: point this PR to the branch of #24 (I can't do immediately, because there is no new commit wrt that one)
> 
> Motivation: having a meaningful diff (i.e. a meaningful view in the _Files changed_ tab)

I have rebased PR to  qibotn_integration branch.

**Comment by liweintu** [comment]
Heads-up:
This PR is branched out from PR #24, that's why it has all the historical commits from branch `qibotn_integration`.  We need to be careful about the two PR branches being merged into `main`.

**Comment by alecandido** [comment]
@liweintu and @Vinitha-balachandran is this completed?

I already had a look to the changes, and I have nothing major to amend.

The only piece that is clearly missing are the extracted docstrings. You could follow the Qibolab example:
https://github.com/qiboteam/qibocal/blob/06457b828f0f8e4c1878266ea48924e6503b15a4/doc/source/conf.py#L126-L131
https://github.com/qiboteam/qibocal/blob/06457b828f0f8e4c1878266ea48924e6503b15a4/doc/source/index.rst?plain=1#L52-L56
(even though I'd `gitignore` the resulting folder, despite Qibocal is not doing it... but maybe it will!)

Other than that, I see that the PR is still draft, so I assume it to be work-in-progress. Is it correct?

**Comment by liweintu** [comment]
> I noticed, and @Vinitha-balachandran changed the base branch.
> 
> The worst it can happen now is that this will be merged in #24, increasing the diff. But it would not be a big deal anyhow, since this is mostly contained in a new folder.

Indeed. Now that PR #35 is planned to be merged into the branch of PR #24 , we'll do PR #35 first. Then, followed by PR #24 . This should pick up all the commits from both branches.

**Comment by alecandido** [comment]
> Indeed. Now that PR #35 is planned to be merged into the branch of PR #24 , we'll do PR #35 first. Then, followed by PR #24 . This should pick up all the commits from both branches.

That's acceptable, but if possible, I'd prioritize #24 first.

The rationale is that this is an addition, so it is not strictly needed for #24, and having #24 in `master` would also improve the testing for the PR in `qibo` itself.

File: .github/workflows/rules.yml
Comment (alecandido) [inline .github/workflows/rules.yml:1]: This workflow is for testing, please restore it to its previous state.

We have dedicated workflows for the docs, and you have to invoke the correct reusable one (you're still calling the test workflow, despite installing only docs dependencies).

You could follow the example of `qibo` itself:
https://github.com/qiboteam/qibo/blob/master/.github/workflows/publish.yml

Comment (Vinitha-balachandran) [reply .github/workflows/rules.yml:1]: My bad, still in the learning stage. Has restored the file.

**Comment by liweintu** [comment]
> That's acceptable, but if possible, I'd prioritize #24 first.
> 
> The rationale is that this is an addition, so it is not strictly needed for #24, and having #24 in `master` would also improve the testing for the PR in `qibo` itself.

Yup, fair enough. We'll go through the remaining points in #24 and finish it up first.

**Comment by Vinitha-balachandran** [comment]
> @liweintu and @Vinitha-balachandran is this completed?
> 
> I already had a look to the changes, and I have nothing major to amend.
> 
> The only piece that is clearly missing are the extracted docstrings. You could follow the Qibolab example: https://github.com/qiboteam/qibocal/blob/06457b828f0f8e4c1878266ea48924e6503b15a4/doc/source/conf.py#L126-L131 https://github.com/qiboteam/qibocal/blob/06457b828f0f8e4c1878266ea48924e6503b15a4/doc/source/index.rst?plain=1#L52-L56 (even though I'd `gitignore` the resulting folder, despite Qibocal is not doing it... but maybe it will!)
> 
> Other than that, I see that the PR is still draft, so I assume it to be work-in-progress. Is it correct?

I have included the missing part.

File: docs/source/index.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+   :maxdepth: 2
+   :caption: Introduction
+   usage
+   api
+
+.. toctree::
+   :maxdepth: 2
+   :caption: Main documentation
+
+    api-reference/modules
```
Comment (liweintu) [inline docs/source/index.rst:60]: We need to put in the right content based on the basic source files available now.

**Review by liweintu — CHANGES_REQUESTED** [review]

**Comment by liweintu** [comment]
> That's acceptable, but if possible, I'd prioritize #24 first.
> 
> The rationale is that this is an addition, so it is not strictly needed for #24, and having #24 in `master` would also improve the testing for the PR in `qibo` itself.

Now that PR #24 has been merged into `main`, I changed the target branch to `main` as per the exact rationale above.

File: docs/source/conf.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+# https://github.com/readthedocs/recommonmark/blob/ddd56e7717e9745f11300059e4268e204138a6b1/docs/conf.py
+# app setup hook
+
+
+def run_apidoc(_):
+    """Extract autodoc directives from package structure."""
+    source = Path(__file__).parent
+    docs_dest = source / "api-reference"
+    package = source.parents[1] / "src" / "qibotn"
+    apidoc.main(["--no-toc", "--module-first", "-o", str(docs_dest), str(package)])
```
[the remark this answers is not in the document]
Comment (liweintu) [reply docs/source/conf.py:73]: Following the qibocal example, added the `setup()` function for Sphinx to call.

File: docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+--------
+
+.. toctree::
+   :maxdepth: 2
+   :caption: Introduction
+   Installation
+   Quickstart
+
+
+
```
[the remark this answers is not in the document]
Comment (liweintu) [reply docs/source/index.rst:60]: Added the usual sections as per the Qibo template.

File: docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+--------
+--------
+
+.. toctree::
+   :maxdepth: 2
+   :caption: Introduction
+   Installation
+   Quickstart
+
+
```
[the remark this answers is not in the document]
Comment (liweintu) [reply docs/source/index.rst:62]: Added the `toctree` section for the generated API.

**Comment by alecandido** [comment]
```
.../qibotn/docs/source/index.rst:46: ERROR: Error in "toctree" directive:
invalid option block.
                                                                                                       
.. toctree::
   :maxdepth: 2
   :caption: Introduction
   Installation
   Quickstart
.../qibotn/docs/source/index.rst:44: ERROR: Document or section may notbegin with a transition.
```

**Comment by liweintu** [comment]
> ```
> .../qibotn/docs/source/index.rst:46: ERROR: Error in "toctree" directive:
> invalid option block.
>                                                                                                        
> .. toctree::
>    :maxdepth: 2
>    :caption: Introduction
>    Installation
>    Quickstart
> .../qibotn/docs/source/index.rst:44: ERROR: Document or section may notbegin with a transition.
> ```

After trying `poetry install --with docs`, I re-produced the error, will follow some reference from `qibo` or `qibocal` to fix it.

**Comment by liweintu** [comment]
> Missing **unused** dependency
> 
> ```
> ❯ make html
> Running Sphinx v5.3.0
> 
> Extension error:
> Could not import extension nbsphinx (exception: No module named 'nbsphinx')
> make: *** [Makefile:20: html] Error 2
> ```
> 
> Please, try to install from scratch with `poetry install --with docs`, and then run `cd docs; make html`, to test the generation is happening correctly.

After re-organising the files into the subfolders as per Qibo document convention, `make html` can work.

However, I encountered a hanging issue of `poetry install` in some GPU Ubuntu machine. In some other machine (e.g. MacBook), `poetry install` works fine. @Tankya2 also experienced this. It seems to be a known issue as per here: python-poetry/poetry#3365. Did you ever encounter this hanging issue? @alecandido

**Comment by scarrazza** [comment]
> @scarrazza could you set the repo to automatically delete merged branches, and reroute existing PRs?

Done.

**Comment by alecandido** [comment]
> It seems to be a known issue as per here: [python-poetry/poetry#3365](https://github.com/python-poetry/poetry/issues/3365). Did you ever encounter this hanging issue? @alecandido

Yes, it was a common problem around Poetry v1.1.2 and similar.
By now (Poetry v1.8 released few days ago, I'm also using v1.7.1), I've never encountered any longer.

If the version of Poetry you're using, I'd suggest upgrading it (with the various methods for installation listed in the docs, e.g. `pipx`, but it's just an example), and deleting the cache (`poetry cache clear PyPI`, or `poetry cache clear <id>`, where you could find `<id>` running `poetry cache list`).

@liweintu

**Review by alecandido — APPROVED** [review]
By now, docs are compiling in a fresh new repo without raising errors nor warnings

**Pull request merged**